File: src/com/app/test/game/ui/AnswerActivity.java
package com.app.test.game.ui;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.TextView;

import com.app.test.R;
import com.app.test.game.bean.Question;
import com.app.test.game.presenter.AnswerPresenter;
import com.app.test.game.source.AnswerType;
import com.app.test.game.view.IdiomView;
import com.app.test.game.view.OpenDoorView;
import com.app.test.game.view.SingleView;
import com.app.test.util.Utils;

/**
 * @author lcx
 * Created at 2020.3.27
 * Describe:
 */
public class AnswerActivity extends Activity {
    private AnswerPresenter answerPresenter;
    private View ivFinish;
    private SingleView singleView;
    private IdiomView idiomView;
    private OpenDoorView openDoorView;
    private TextView tvProgress;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_answer_question);
        initView();
        initData();
    }

    private void initView() {
        ivFinish = findViewById(R.id.iv_finish);
        singleView = findViewById(R.id.sgv_single);
        idiomView = findViewById(R.id.sgv_proverb);
        openDoorView = findViewById(R.id.godv_layout);
        tvProgress = findViewById(R.id.tv_answer_progress);
        setViewListener();
    }

    private void initData() {
        Intent intent = getIntent();
        if (Utils.isEmpty(intent)) {
            finish();
            return;
        }
        getPresenter().getCheckpointList();
    }

    private AnswerPresenter getPresenter() {
        if (answerPresenter == null)
            answerPresenter = new AnswerPresenter(this);
        return answerPresenter;
    }


    private void setViewListener() {
        ivFinish.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
        idiomView.setAnswerListener(getPresenter());
        singleView.setAnswerListener(getPresenter());
    }


    public void dispatchQuestion(final int index, final Question question) {
        if (Utils.isEmpty(openDoorView)) {
            _dispatchQuestion(index, question);
            return;
        }
        openDoorView.startCloseAnim();
        openDoorView.setOpenDoorListen(new OpenDoorView.OpenDoorListen() {
            @Override
            public void openAnimStart() {
                _dispatchQuestion(index, question);
            }
        });
    }


    public void _dispatchQuestion(int index, Question question) {
        if (checkIllegal(index, question)) {
            return;
        }
        int questionType = question.getQuestionType();
        singleView.setVisibility(View.GONE);
        idiomView.setVisibility(View.GONE);
        switch (questionType) {
            case AnswerType.NORMAL:
                singleView.setVisibility(View.VISIBLE);
                singleView.onNextQuestion(index, question);
                break;
            case AnswerType.PROVER:
                idiomView.setVisibility(View.VISIBLE);
                idiomView.onNextQuestion(index, question);
                break;
        }
    }

    public void changeIndex(int index, int amount) {
        if (Utils.isEmpty(tvProgress) || index + 1 > amount) {
            return;
        }
        tvProgress.setText(String.format("本轮答题:%1$s/%2$s题", String.valueOf(index + 1), String.valueOf(amount)));
    }


    @Override
    protected void onDestroy() {
        if (!Utils.isEmpty(singleView))
            singleView.onDestroy();
        if (!Utils.isEmpty(idiomView))
            idiomView.onDestroy();
        super.onDestroy();
    }

    private boolean checkIllegal(int index, Question question) {
        return Utils.isEmpty(singleView) || Utils.isEmpty(idiomView) || index < 0 || Utils.isEmpty(question) || question.getQuestionType() == AnswerType.NONE;
    }
}

File: src/com/app/test/game/ui/AnswerActivity.kt
package com.app.test.game.ui

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.widget.TextView
import com.app.test.R
import com.app.test.game.bean.Question
import com.app.test.game.presenter.AnswerPresenter
import com.app.test.game.source.AnswerType
import com.app.test.game.view.IdiomView
import com.app.test.game.view.OpenDoorView
import com.app.test.game.view.OpenDoorView.OpenDoorListen
import com.app.test.game.view.SingleView
import com.app.test.util.Utils

/**
 * @author lcx
 * Created at 2020.3.27
 * Describe:
 */
class AnswerActivity : Activity() {
    private val answerPresenter: AnswerPresenter by lazy { AnswerPresenter(this) }
    private lateinit var ivFinish: View
    private lateinit var singleView: SingleView
    private lateinit var idiomView: IdiomView
    private lateinit var openDoorView: OpenDoorView
    private lateinit var tvProgress: TextView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_answer_question)
        initView()
        initData()
    }

    private fun initView() {
        ivFinish = findViewById(R.id.iv_finish)
        singleView = findViewById(R.id.sgv_single)
        idiomView = findViewById(R.id.sgv_proverb)
        openDoorView = findViewById(R.id.godv_layout)
        tvProgress = findViewById(R.id.tv_answer_progress)
        setViewListener()
    }

    private fun initData() {
        val intent = intent
        if (Utils.isEmpty(intent)) {
            finish()
            return
        }
        answerPresenter.checkpointList
    }


    private fun setViewListener() {
        ivFinish.setOnClickListener { finish() }
        idiomView.setAnswerListener(answerPresenter)
        singleView.setAnswerListener(answerPresenter)
    }

    fun dispatchQuestion(index: Int, question: Question) {
        if (Utils.isEmpty(openDoorView)) {
            _dispatchQuestion(index, question)
            return
        }
        openDoorView.startCloseAnim()
        openDoorView.setOpenDoorListen(object : OpenDoorListen() {
            override fun openAnimStart() {
                _dispatchQuestion(index, question)
            }
        })
    }

    fun _dispatchQuestion(index: Int, question: Question) {
        if (checkIllegal(index, question)) {
            return
        }
        val questionType = question.questionType
        singleView.visibility = View.GONE
        idiomView.visibility = View.GONE
        when (questionType) {
            AnswerType.NORMAL -> {
                singleView.visibility = View.VISIBLE
                singleView.onNextQuestion(index, question)
            }
            AnswerType.PROVER -> {
                idiomView.visibility = View.VISIBLE
                idiomView.onNextQuestion(index, question)
            }
        }
    }

    fun changeIndex(index: Int, amount: Int) {
        if (index + 1 > amount) {
            return
        }
        tvProgress.text = String.format("本轮答题:%1\$s/%2\$s题", (index + 1).toString(), amount.toString())
    }

    override fun onDestroy() {
        singleView.onDestroy()
        idiomView.onDestroy()
        super.onDestroy()
    }

    private fun checkIllegal(index: Int, question: Question): Boolean {
        return Utils.isEmpty(singleView) || Utils.isEmpty(idiomView) || index < 0 || Utils.isEmpty(question) || question.questionType == AnswerType.NONE
    }
}